lib/Yield/Support/GraphLayout/SugiyamaStyle/GraphPreparation.cpp

```diff
@@ -47,32 +48,63 @@ static SelfLoopContainer extractSelfLoops(InternalGraph &Graph) {
 /// real entry point.
 static void
 ensureSingleEntry(InternalGraph &Graph, RankContainer *MaybeRanks = nullptr) {
-  auto EntryNodes = entryPoints(&Graph);
-  revng_assert(!EntryNodes.empty());
+  auto EntryPoints = entryPoints(&Graph);
+  revng_assert(!EntryPoints.empty());
 
-  if (EntryNodes.size() == 1) {
+  if (EntryPoints.size() == 1) {
     // If there's only a single entry point, make sure it's set.
-    Graph.setEntryNode(EntryNodes.front());
+    Graph.setEntryNode(EntryPoints.front());
   } else {
     // If there's more than one, add a new virtual node with all the real
     // entry nodes as its direct successors.
-
-    // BUT if the currently set entry node is already virtual, remove it first:
-    // this prevents the possibility of chaining virtual entry nodes when this
-    // function is invoked on a slightly-modified graph multiple times.
-    if (Graph.getEntryNode() != nullptr) {
-      if (Graph.getEntryNode()->IsVirtual) {
-        Graph.removeNode(Graph.getEntryNode());
-        if (MaybeRanks != nullptr)
-          MaybeRanks->erase(Graph.getEntryNode());
+    InternalNode *EntryNode = Graph.getEntryNode();
+    if (EntryNode != nullptr) {
+      // If there is already an entry node present for the graph,
+      if (EntryNode->IsVirtual) {
+        // and it's virtual, reuse it,
+      } else {
+        // otherwise, ignore it.
+        EntryNode = nullptr;
       }
```

Comment:
I don't this can ever happen. If it could happen we'd have a graph such that:
- `entryPoints(&Graph)` returns many entry points
- `Graph.getEntryNode` returns a single entry point that is not virtual

Intuitively, if this happens there must be something very wrong, don't you think? Have you seen cases where it happens in practice? If so, can you share them? Otherwise I'd just assert that whenever you have many `EntryPoints` then `EntryNode == nullptr or EntryNode->IsVirtual`.

Reply:
Since we split back-edges into `V`-shapes, they add new entry points. Hence the trickery and the need to "fixup" the entry point.

Reply:
Ok then, please document this.

Reply:
Or better, can you assert that this happens only for the nasty V-shaped ones?

Reply:
Actually, that's a really good idea!

Reply:
But wait, now that I actually think about it, what is somebody passes a graph with the entry point already set into the layouter, except it's not the only entry point. That case would also trigger this branch. Should we also assert that the entry point is not set when we get the hold of the graph? Or maybe quietly override it?

Reply:
The more I think about it, the more I'm starting to feel like this should be left as is.

Consider the broad picture:
```
1. find all the entry nodes.
2. if there's only one, set it as "official" and get out.
3. otherwise, check if we already have an "official" entry point set.
5. if there is one, and it's marked as "artificial", just update it.
6. if there is one, but it's not marked as "artificial", ignore it and make a new one.
7. if there's none - also make a new one.
8. fill the successors of the new node and get out.
```

which implicitly makes sense to me - we make as few assumptions about the graph as we can, while also making sure the algorithm still works correctly when it's called multiple times on a graph that was modified in-between (mainly, new entry points appeared).

Is it that advantageous to make sure the newly added nodes are always V-shaped? So much so as to compromise how general the check is?

Reply:
Not sure. I'm pretty concerned about 2. and 6.

In both cases you might be getting a graph where the "official" entry point has been set from outside but you ignore it. This is very very weird from a user's standpoint. Imagine if the user is trying to set the "official" entry point from outside, so that the routing starts from there, but we actually ignore it and route the way we like it considering another entry point. As a user I would be pissed off.

I'd be much happier to assert that we never get a graph from the user that has an "official" entry point already set.

Then if it happens we can detect it and decide what to do.

But in principle, until I don't see such a case, I'd like to rule this out and make this as part of the "contract" of the API: you just don't send in a graph with an "official" entry point already set, or the implementation will assert.

Or you're saying that you already have in mind a situation where we might wanna do exactly that?

Reply:
It's not that your proposal can't work, but why bothering if we don't even envision users to provide us an entry point that is decided by them?

Let's just decide that the layouter API has a contract: the place and route algorithm works **only** with an entry point that **we** decide. If someone tries to be a smartass and sets the entry point from outside, they will just get an assertion in their face and learn to respect the API's contract.

I don't want to be stuck supporting a half-baked counterintuitive scenario unless we find at least a real use case for which this is a real game changer. Do you have one of these uses cases in mind? Or you're just speculating?

Reply:
As per the mattermost discussion, I'm closing this thread.